app/src/backend/ActivityService.ts

```diff
@@ -21,9 +21,241 @@ import type { Transaction } from "sequelize";
 
 type GasValueInput = Omit<GasValueCreationAttributes, "id"> & {
   emissionsFactor?: Omit<EmissionsFactorAttributes, "id">;
+  id?: string;
 };
 
 export default class ActivityService {
+  private static async updateDataSource({
+    activityValue,
+    dataSourceParams,
+    transaction,
+  }: {
+    activityValue: ActivityValue;
+    dataSourceParams: Omit<DataSourceAttributes, "datasourceId"> | undefined;
+    transaction: Transaction;
+  }): Promise<string> {
+    if (activityValue.datasourceId) {
+      const dataSource = await db.models.DataSource.findOne({
+        where: { datasourceId: activityValue.datasourceId },
+        transaction,
+      });
+      if (!dataSource) {
+        throw new createHttpError.NotFound(
+          "Data source for ActivityValue not found",
+        );
+      }
+      Object.assign(dataSource, dataSourceParams);
+      await dataSource?.save({ transaction });
+      return dataSource.datasourceId;
+    } else {
+      const dataSource = await db.models.DataSource.create(
+        {
+          ...dataSourceParams,
+          datasourceId: randomUUID(),
+        },
+        { transaction },
+      );
+      return dataSource.datasourceId;
+    }
+  }
+
+  private static async updateInventory({
+    activityValue,
+    inventoryValueParams,
+  }: {
+    activityValue: ActivityValue;
+    inventoryValueParams?: Omit<InventoryValueAttributes, "id"> | undefined;
```

Comment:
```suggestion
    inventoryValueParams?: InventoryValueAttributes | undefined;
```
You need the id in order to know which one you need to update.

Reply:
I guess you can also get it from the activity value like you do below...